Modeler : operators ceil and floor [ANT-4295]

guilpier-code opened the pull request:
Adding the **ceil** and **floor** operators.

You should know that the current PR is followed by 2 more PRs (in a sense based on this one and on one another) : 
- [PR 3329](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/3329), whose purpose is to simplify the forbidden nodes part
- [PR 3335](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/3335), whose purpose is to add features on **ceil** and **floor** operators that were forgotten / not yet done in the current PR (especially **ceil** operator and unit tests).

Commit:
operators ceil and floor : mainly make code more clear

Commit:
operators ceil and floor : fix typo

Commit:
operators ceil and floor : renaming

Commit:
operators ceil and floor : mainly remove unnecessary use of a fixture - part 1

Commit:
operators ceil and floor : remove fixture ExpressionToNodeConvertorEmptyModel, useless

Commit:
operators ceil and floor : add a fixture holding a registry

Commit:
operators ceil and floor : forgot a use for the registry holder fixture

Commit:
cleaning before ceil and floor : remove a non relevant method

Commit:
Revert "cleaning before ceil and floor : remove a non relevant method"

This reverts commit 7eb46071159d2f2b185920f99d59d7fd747b5aef.

Commit:
cleaning before ceil and floor : remove a code duplication

Commit:
ceil and floor : Convertor visitor >  add an floor case to FunctionNode + simplify tests (by adding fixtures)

Commit:
ceil and floor : Convertor visitor >  make unit tests a fraction more clear

Commit:
ceil and floor : Convertor visitor >  in unit tests, remove useless namespace prefixing

Commit:
floor operator : Convertor visitor > add a first (nominal) case

Commit:
floor operator : Convertor visitor > add 2 more unit tests about wrong use of floor(...)

Commit:
floor operator : Convertor visitor > add a test on wrong use of floor (passing it a variable)

guilpier-code reviewed (commented):
blabla

guilpier-code commented on src/expressions/include/antares/expressions/nodes/FunctionNode.h:
**FunctionNodeTypeToString** : this function should only be defined in **.cpp** source file, as it's called inside the class only.

guilpier-code replied:
Was moved to corresponding  cpp file in PR [3335](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/3335)

guilpier-code commented on src/tests/src/io/yml-importers/testConvertorVisitor.cpp:
```diff
@@ -661,181 +562,143 @@ BOOST_AUTO_TEST_CASE(ValidMaxExpression)
+                          checkMessage(err_msg));
+}
+
+BOOST_FIXTURE_TEST_CASE(floor_operator___nominal_case, SupplyModelForFunctionalOperator)
```
Here we add several unit tests about wrong use of **floor** operator. 
For each wrong use, an exception is raised.
Do we want to use the **NodeChecker** as well (see **modelConverter.cpp**) ? Wouldn't it be a double check or a kind of duplication ?

a-zakir replied:
you shouldn't check in the convertvisitor, his responsability is only to convert antlr tree.

a-zakir replied:
for now it is hard coded that floor expects either param or literal node. why would user do floor(2.3) when he can simply write 2 ?

guilpier-code replied:
Checks done in **convert visitor** turn out to be useful.
Indeed, there are 2 kind of prohibition about floor operators (an ceil) : 
- we have to control the number of arguments of floor (less than one and more than one is not accepted)
- we have to control the type of arguments floor can take and in which conditions.

First checks are done in the **convert visitor**, and I can see a better place to do them.

Commit:
floor operator : Eval visitor > small simplifications

Commit:
floor operator : Evaluation visitor > add the floor op + make code a bit more readable.

Commit:
floor operator : Evaluation visitor > add unit tests

Commit:
[skip ci] floor operator > Linearity visitor : adding a comment for uselessness of this class

Commit:
floor : Print visitor > make a function member a free function

flomnes commented on src/io/inputs/model-converter/convertorVisitor.cpp:
```diff
@@ -482,12 +484,51 @@ std::any ConvertorVisitor::handleMin(ExprParser::ArgListContext* context)
+    // Try to create a literal node
+    try
+    {
+        const double arg_as_d = std::stod(arg_as_str);
```
1. Use `std::from_chars` instead
2. In the case of a literal, the `std::floor` function needs to be applied to the result of the `std::string` -> `double` conversion.

guilpier-code replied:
Visiting floor operator (see **ConvertorVisitor::visitFloor**) is now much more simple (as it should be), and requires use of neither `std::floor` nor `std::stod`

Commit:
floor : Print visitor > simplifications

Commit:
Merge branch 'develop' into feature/ceil-and-floor--ant-4295

Commit:
floor operator : fix a compilation error, due to previous merge

Commit:
function operators > print visitor : simplifications

Commit:
function operators > print visitor : more simplifications

Commit:
function operators > visitors : rename and refactor visiting function arguments

Commit:
function operators > visitors : simplifications

Commit:
function operators > visitor : renaming a ill-named header file

Commit:
function operators > print visitor : fix a unit test

guilpier-code commented on src/tests/src/expressions/test_PrintAndEvalVisitors.cpp:
```diff
@@ -1462,7 +1460,7 @@ BOOST_FIXTURE_TEST_CASE(PrintDualNode, MyDummyFixture)
                                       create<LiteralNode>(0));
     PrintVisitor printVisitor;
-    BOOST_CHECK(printVisitor.dispatch(dual) == "dual(constraint)");
+    BOOST_CHECK(printVisitor.dispatch(dual) == "dual(constraint, 0.000000)");
```
**Careful here** : a test check was changed. Is this was we want ?

a-zakir replied:
no

guilpier-code replied:
But a node of type `FunctionNodeType::dual` takes 2 arguments (a constraint and an index), and the **PrintVisitor** should print both. 
So what do we do ?
**dual(contraint)[index]** ?

guilpier-code replied:
I changes dual node print into following format **dual(contraint)[index]**.
But I did this in [PR 3335](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/3335)

Commit:
function floor operators > print visitor : adding  unit tests

Commit:
function operators : move visitFunctionArgs to parent node visitor

Commit:
function operators > visitor : move a member function definition outside the class

Commit:
function operators > visitor : fix compilation failure + renaming

Commit:
function operators > print visitor : simplifying

Commit:
function floor operators > variability visitor : adding floor + adding unit tests

Commit:
function operators > visitors : fix a compilation error on Ubuntu

Commit:
function operators > visitors : fix a compilation error on Ubuntu (second attempt)

Commit:
function floor operators > ReadLinearExpressionVisitor : code formatting

Commit:
function floor operators > ReadLinearExpressionVisitor : renaming and cleaning a bit

Commit:
function floor operators >  ReadLinearExpressionVisitor : code formatting

flomnes commented on src/expressions/visitors/EvalVisitor.cpp:
```diff
@@ -244,6 +244,20 @@ EvaluationResult EvalVisitor::handlePow(const Nodes::FunctionNode* node)
+{
+    auto* floor_arg = node->getOperands()[0];
+    return dispatch(floor_arg).evaluateUnaryOperation(Floor());
+}
```
It's not necessary to construct a temporary `Floor` object here

guilpier-code replied:
ok, but it's a bit more clear to give the floor argument a name.

flomnes commented on src/io/inputs/model-converter/convertorVisitor.cpp:
```diff
@@ -482,12 +484,51 @@ std::any ConvertorVisitor::handleMin(ExprParser::ArgListContext* context)
     auto* arglist = context->argList();
 
-    if (!arglist)
+    if (!arglist || !arglist->expr().size())
```
`!arglist || arglist->expr().empty()`

guilpier-code replied:
Done

Commit:
function floor operators >  visitor : more readability

a-zakir commented on src/io/inputs/model-converter/convertorVisitor.cpp:
```diff
@@ -482,12 +484,51 @@ std::any ConvertorVisitor::handleMin(ExprParser::ArgListContext* context)
+        return registry_.create<LiteralNode>(arg_as_d);
+    }
+    catch (const std::invalid_argument&)
+    {
```
Why are combinations of parameters and literals not handled in the current implementation?
For example:

floor(param * 3)
floor(param1 * param2)
floor(param * lit1 + lit2)
Note:
Since floor is a non-linear function, the following does not hold:
floor(literal * param) != literal * floor(param)

a-zakir replied:
also floor can be applied to variable in extra-outputs

a-zakir commented on src/tests/src/expressions/test_PrintAndEvalVisitors.cpp:
```diff
@@ -1319,22 +1317,22 @@ BOOST_FIXTURE_TEST_CASE(functionNode_pow, MyDummyFixture)
     const auto printed = printVisitor.dispatch(pow);
 
-    BOOST_CHECK_EQUAL(printed, "22.000000^(2.000000)");
+    BOOST_CHECK_EQUAL(printed, "pow(22.000000, 2.000000)");
```
why do you change this print ?

guilpier-code replied:
I can reverse this.
But I did it because I tried to replace these help functions : 
- **handleReducedCost**, 
- **handleDual**, 
- **ProcessOtherFunction**, 
- **handlePow**,
- **handleFloor**

with a unique (more generic) function. and it was easier to use 'pow' (as a result of `node->typeToString()`)

a-zakir replied:
Ok , i understand. Pow is unique among function nodes. It is the only one with a symbol. The printvisitor tries to reflect what the user may enter.

Commit:
function floor operator > Read Linear Expression Visitor :add floor operator + unit test + readability + simplifications

- make the change for floor operator
- adding 1 unit test
- changes for a bit more readability
- changes for a bit more simplicity

Commit:
function floor operator >  Read Linear Expression Visitor : add a unit test on a literal

Commit:
floor operator : Convertor visitor > change due to review

Commit:
floor op : Convertor visitor > handling literal node

Commit:
floor op : Convertor visitor > fix visitFloor function

flomnes commented on src/solver/optim-model-filler/ReadLinearExpressionVisitor.cpp:
```diff
@@ -312,31 +294,45 @@ Optimization::TimeDependentLinearExpression ReadLinearExpressionVisitor::handleP
+    auto expressions = dispatch(node->getOperands()[0]);
+    for (auto& expression: expressions)
+    {
+        // gp : should be removed when arg of floor operator applies to extra output as well.
```
https://github.com/AntaresSimulatorTeam/Antares_Simulator/blob/3a6ce6438168c953858faa9d07b75e46d4bc4d40/src/io/outputs/SimulationTableGenerator.cpp#L265-L280

This visitor isn't used to evaluate `ExtraOutputs`, we use `EvalVisitor` for that.

guilpier-code replied:
ok, but why this remark ?

flomnes replied:
You can remove the comment

guilpier-code replied:
Comment was removed in PR [3335](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/3335)
See file **src/solver/optim-model-filler/ReadLinearExpressionVisitor.cpp**

guilpier-code replied:
Was removed in [PR 3335](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/3335)

flomnes commented on src/tests/src/solver/optim-model-filler/test_readLinearExpressionVisitor.cpp:
```diff
@@ -161,6 +161,33 @@ BOOST_FIXTURE_TEST_CASE(visit__min_literal__param, VisitorFixture<ReadLinearExpr
     BOOST_CHECK_EQUAL(linear_expression[0].size(), 0);
 }
 
+BOOST_FIXTURE_TEST_CASE(linear_expr_from_floor_of_a_literal,
+                        VisitorFixture<ReadLinearExpressionVisitor>)
```
You may also add `floor(4.5)*x`, where `x` is a variable node.

guilpier-code replied:
Done in [PR 3335](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/3335)

Commit:
Convertor visitor > small clarifications

Commit:
Convertor visitor > small clarifications

Commit:
Convertor visitor > remove code duplication

Commit:
floor : Print visitor > make review easier

Commit:
print visitor : simplification

Commit:
Eval visitor : renaming

Commit:
Eval visitor : simplification

Commit:
Eval visitor : simplification

guilpier-code committed:
Modeler : simplify forbidden nodes (#3329)

This PR is the 2nd PR out of 3 PR about adding operators **ceil** and
**floor** :
- [PR
3316](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/3316)
(the main work on adding these operators)
- [PR
3329](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/3329),
(the current one)
- [PR
3335](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/3335)

It intends to simplify and make more readable code about forbidden
nodes.

**What was done** : some PR comments were left below. Please read them.

---------

Co-authored-by: guilpierre <guillaume.pierre_externe@rte-france.com>

guilpier-code committed:
ceil and floor operator : add missing features (#3335)

This PR is the 3rd PR out of 3 PR about adding operators **ceil** and
**floor** :
- [PR
3316](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/3316)
(the main work on adding these operators)
- [PR
3329](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/3329),
its purpose is to simplify the forbidden nodes part
- [PR
3335](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/3335)
(the current PR)

This last PR's purpose is to add features on **ceil** and **floor**
operators that were forgotten / not yet done in the current PR
(especially **ceil** operator and unit tests).

Note : 
- [x] Linearity visitor turns out to be useless, so it was removed (unit
tests as well)
- [ ] what else ?

---------

Co-authored-by: guilpierre <guillaume.pierre_externe@rte-france.com>
Co-authored-by: Florian OMNES <florian.omnes@rte-france.com>

flomnes committed:
Merge branch 'develop' into feature/ceil-and-floor--ant-4295

flomnes committed:
Add missing tests
